text/0000-imply-option.md

```diff
@@ -0,0 +1,139 @@
+But the `else` branch is required for returning `None` value if `x == 0` evaluates to false.
+Fortunately Rusts `Option` type has functionality get rid of the unecessary code:
+```
+    let x = 0;
+    
+    Option::on_pred(x == 0, x)
+```
+This code performs the exact same function as our original `if` statement however our code is compressed to a single line and our intentions are just as clear.
+Have you spotted the possible issue with this solution introduces however? What about this code:
+```
+    Option::on_pred(false, foo())
+```
+The above line of code will always return `None` and always throw away the result of `foo()` wasting our precious computing power every time our code needs to return `None`.
+Rust has thought ahead of this problem though:
+```
+    Option::lazy_pred(false, foo)
+```
+`Option`s `lazy_pred` function leverages lazy evaluation by taking a function pointer as its second argument. If its first argument evaluates to `true` it will return `Some(foo())` but if its first argument is `false` it returns `None` without having to run `foo`. This solves the problem presented in our earlier example without sacrificing the advantages it gave us.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This addition is in essence an implementation of the "implies" or "->" operation from predicate logic.
+In predicate logic, for those unfamiliar, the "->" operation has the following truth table:
+```
+    | x | y | x -> y |
+    | F | F | T      |
+    | F | T | T      |
+    | T | F | F      |
+    | T | T | T      |
+```
+The Rust addition I am suggesting can be encapsulated as "If `x` is `true`, I care about the value of `y`; else I do not care about the value of `y`." or:
+```
+    | x | x -> y |
+    | F | None   |
+    | T | Some(y)|
+```
+My initial proposal for how this addition could be implemented is:
+```
+    impl<T: Sized> Option<T> {
+        /// A straight forward implementation of the `implies` operation for predicate logic.
+        fn on_pred(pred: bool, value: T) -> Self {
+            if pred {
+                Some(value)
+            } else {
+                None
+            }
+        }
+        /// A lazy implementation of the `implies` operation when necessary.
+        fn lazy_pred<F>(pred: bool, func: F) -> Self
+            where F: FnOnce() -> T {
+            if pred {
+                Some(func())
+            } else {
+                None
+            }
+        }
+    }
+```
+This implementation covers the use cases I've proposed in my earlier examples and any others of similar form without any external dependancies; this should make the implementation stable as Rust continues to develope.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is a functionality which has functional programming and monads in mind with its design and this may make it another stepping stone to be learned for programmers which are new to Rust or functional programming concepts.
```

Comment:
Rust already has a lot of FP concepts and actual monadic bind in `Option::and_then`, `Iterator::flat_map` in libcore - I don't see why adding a small addition like this would be anything new FP-wise.

Reply:
I'm not saying it would be anything new, rather it would be just one more little thing which needs to be learned when considering new programmers.